MLPAutoCompleteTextField/MLPAutoCompleteTextField.m

```diff
@@ -911,45 +911,109 @@ - (NSAttributedString *)boldedString:(NSString *)string withSubstrings:(NSString
 
-                [operation performSelector:@selector(didReceiveSuggestions:) withObject:suggestions];
-                waitingForSuggestions = NO;
-            }];
+    if (oldStateKeyPath) [self willChangeValueForKey:oldStateKeyPath];
+    if (newStateKeyPath) [self willChangeValueForKey:newStateKeyPath];
 
-            while(waitingForSuggestions) {
-                [NSThread sleepForTimeInterval:250];
+    _state = aState;
 
-                if(self.isCancelled) {
-                    return;
-                }
-            }
+    if (oldStateKeyPath) [self didChangeValueForKey:oldStateKeyPath];
+    if (newStateKeyPath) [self didChangeValueForKey:newStateKeyPath];
+}
 
-        } else if ([self.dataSource respondsToSelector:@selector(autoCompleteTextField:possibleCompletionsForString:)]) {
-            NSArray *results = [self.dataSource autoCompleteTextField:self.textField possibleCompletionsForString:self.incompleteString];
+- (BOOL)isExecuting
+{
+    return self.state == MLPAutoCompleteOperationStateExecuting;
+}
 
-            if(!self.isCancelled) {
-                [self didReceiveSuggestions:results];
-            }
+- (BOOL)isFinished
+{
+    return self.state == MLPAutoCompleteOperationStateFinished;
+}
+
+- (BOOL)isAsynchronous
+{
+    return self.canPerformAsync;
+}
+
+- (BOOL)canPerformAsync
+{
+    return [self.dataSource respondsToSelector:@selector(autoCompleteTextField:possibleCompletionsForString:completionHandler:)];
+}
+
+- (BOOL)canPerformSync
+{
+    return [self.dataSource respondsToSelector:@selector(autoCompleteTextField:possibleCompletionsForString:)];
+}
 
-        } else {
-            NSAssert(0, @"An autocomplete datasource must implement either autoCompleteTextField:possibleCompletionsForString: or "
-                "autoCompleteTextField:possibleCompletionsForString:completionHandler:");
+- (void)start
+{
+    if (self.isCancelled) {
+        self.state = MLPAutoCompleteOperationStateFinished;
+        return;
+    }
+
+    if (self.canPerformAsync == NO && self.canPerformSync == NO) {
+        NSAssert(0, @"An autocomplete datasource must implement either autoCompleteTextField:possibleCompletionsForString: or "
+                 "autoCompleteTextField:possibleCompletionsForString:completionHandler:");
+        self.state = MLPAutoCompleteOperationStateFinished;
+        return;
+    }
+
+    self.state = MLPAutoCompleteOperationStateExecuting;
+    [self main];
+}
+
+- (void)main
+{
+    if (self.isCancelled) {
+        self.state = MLPAutoCompleteOperationStateFinished;
+        return;
+    }
+
+    if (self.isAsynchronous) {
+        __weak typeof(self) weakSelf = self;
+        [self.dataSource autoCompleteTextField:self.textField possibleCompletionsForString:self.incompleteString
+                             completionHandler:^(NSArray *suggestions) {
+                                 [weakSelf performSelector:@selector(didReceiveSuggestions:) withObject:suggestions];
```

Comment:
In this code path, if the operation is canceled, it calls `didReceiveSuggestions:`, but...